File: guests/haskell/radio.hs
{-# LANGUAGE OverloadedStrings #-}

-- The radio scene, Haskell port — guests/rust/radio.rs,
-- tools/scenes/radio.steps.

import Data.Text (Text)
import KayaApp

options :: [Text]
options = ["Small", "Medium", "Large"]

main :: IO ()
main = kayaMain $ \app -> do
  _ <- buildTx app $ do
    window primary [WTitle "radio"]
    size <- signalText "size: Small"

    let onPick index =
          submitTx app $
            writeSignal size ("size: " <> options !! index)

    root <-
      column
        []
        [ radioOn options 0 onPick,
          labelBound size -- label#0
        ]
    mount root
    return ()
  return ()

File: guests/rust/radio.rs
//! The radio conformance scene: the choice contract (see select.rs) in
//! its inline presentation. The contract is tools/scenes/radio.steps.

const OPTIONS: [&str; 3] = ["Small", "Medium", "Large"];

#[derive(Clone)]
enum Msg {
    Picked(usize),
}

pub(crate) fn app(ctx: kaya::AppCtx) {
    let msgs = kaya::Messages::new();
    let size = ctx.apply(|tx| {
        tx.window(kaya::DEFAULT_WINDOW).title("radio");
        let size = tx.signal("size: Small");
        let root = tx
            .column(|tx| {
                let group = tx.radio(&OPTIONS, 0).id();
                msgs.on_select(group, Msg::Picked);
                tx.label(size);
            })
            .id();
        tx.mount(root);
        size
    });

    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            Msg::Picked(index) => {
                ctx.apply(|tx| {
                    tx.write(size, format!("size: {}", OPTIONS[index]));
                });
            }
        }
    }
}

fn main() {
    kaya::run(app)
}
